okhttp/src/jvmMain/kotlin/okhttp3/Request.kt

```diff
@@ -32,18 +32,52 @@ import okhttp3.internal.commonPut
 import okhttp3.internal.commonRemoveHeader
 import okhttp3.internal.toImmutableMap
 
-actual class Request internal constructor(
-  @get:JvmName("url") actual val url: HttpUrl,
-  @get:JvmName("method") actual val method: String,
-  @get:JvmName("headers") actual val headers: Headers,
-  @get:JvmName("body") actual val body: RequestBody?,
-  internal val tags: Map<Class<*>, Any>
-) {
+actual class Request internal actual constructor(builder: Builder) {
+  @get:JvmName("url")
+  actual val url: HttpUrl = checkNotNull(builder.url) { "url == null" }
+
+  @get:JvmName("method")
+  actual val method: String = builder.method
+
+  @get:JvmName("headers")
+  actual val headers: Headers = builder.headers.build()
+
+  @get:JvmName("body")
+  actual val body: RequestBody? = builder.body
+
+  internal val tags: Map<Class<*>, Any> = builder.tags.toImmutableMap()
+
   internal actual var lazyCacheControl: CacheControl? = null
 
   actual val isHttps: Boolean
     get() = url.isHttps
 
+  /**
+   * Constructs a new request.
+   *
+   * Use [Builder] for more fluent construction, including helper methods for various HTTP methods.
+   *
+   * @param method defaults to "GET" if [body] is null, and "POST" otherwise.
+   */
+  constructor(
+    url: HttpUrl,
+    headers: Headers = Headers.headersOf(),
+    method: String = "\u0000", // Sentinel value chooses based on what the body is.
```

Comment:
Cute, was a bit against it at first, but it hits the sweet spot for usage.

Reply:
It also removes the comment I came here to make, that HttpMethods should be public API, since you need to pass in here.  But not an issue since this can be left out mostly.

Reply:
I considered `null` as the sentinel, but that lets users pass null which I don’t like.

I considered making the `when` expression the default value, but that puts the parameters in a surprising order.

This is ugly, but users shouldn’t see it.

Reply:
Ooooh think of the poor C programmers, strlen on this guy isn't fun.